Subject: Add overall CMake changes for PHP-8.4 branch

- EditorConfig adjustments for code style support in editors and IDEs.
- CMake build system files added to .gitignore.
- CMake added to run-tests.php for info about failing tests and checks.
- CMakeLists.txt template and modifications added to ext/skeleton.
---
 .editorconfig              |  2 +-
 .gitignore                 | 14 ++++++++++++++
 ext/ext_skel.php           |  1 +
 ext/skeleton/.gitignore.in | 14 ++++++++++++++
 run-tests.php              | 24 ++++++++++++++++++------
 5 files changed, 48 insertions(+), 7 deletions(-)

diff --git a/.editorconfig b/.editorconfig
index 7911bf8490..0ca142015f 100644
--- a/.editorconfig
+++ b/.editorconfig
@@ -17,7 +17,7 @@ indent_style             = tab
 indent_size              = 4
 indent_style             = space
 
-[*.{ac,m4,sh,yml}]
+[{CMakeLists.txt,*.{ac,cmake,json,m4,sh,yml}}]
 indent_size              = 2
 indent_style             = space
 
diff --git a/.gitignore b/.gitignore
index cf11cd0ac2..82274c860d 100644
--- a/.gitignore
+++ b/.gitignore
@@ -296,3 +296,17 @@ tmp-php.ini
 !/ext/dom/lexbor/patches/*.patch
 !/ext/pcre/pcre2lib/config.h
 !/win32/build/Makefile
+
+# ------------------------------------------------------------------------------
+# CMake build system files
+# ------------------------------------------------------------------------------
+/_deps/
+/.ninja*
+/build.ninja
+/CMakeUserPresets.json
+/CTestTestfile.cmake
+/profile.json
+/Testing/
+cmake_install.cmake
+CMakeCache.txt
+CMakeFiles/
diff --git a/ext/ext_skel.php b/ext/ext_skel.php
index ae7a3a987c..648e3728b8 100755
--- a/ext/ext_skel.php
+++ b/ext/ext_skel.php
@@ -302,6 +302,7 @@ function copy_config_scripts() {
         $files[] = 'config.w32';
     }
 
+    $files[] = 'CMakeLists.txt';
     $files[] = '.gitignore';
 
     foreach($files as $config_script) {
diff --git a/ext/skeleton/.gitignore.in b/ext/skeleton/.gitignore.in
index ae434fef97..3b25e9e131 100644
--- a/ext/skeleton/.gitignore.in
+++ b/ext/skeleton/.gitignore.in
@@ -39,3 +39,17 @@ tests/**/*.sh
 tests/**/*.db
 tests/**/*.mem
 tmp-php.ini
+
+# ------------------------------------------------------------------------------
+# CMake build system files
+# ------------------------------------------------------------------------------
+/_deps/
+/.ninja*
+/build.ninja
+/CMakeUserPresets.json
+/CTestTestfile.cmake
+/profile.json
+/Testing/
+cmake_install.cmake
+CMakeCache.txt
+CMakeFiles/
diff --git a/run-tests.php b/run-tests.php
index 5837a1a850..8665548e34 100755
--- a/run-tests.php
+++ b/run-tests.php
@@ -854,10 +854,12 @@ function write_information(array $user_tests, $phpdbg): void
         <?php
         $exts = get_loaded_extensions();
         $ext_dir = ini_get('extension_dir');
-        foreach (scandir($ext_dir) as $file) {
-            if (preg_match('/^(?:php_)?([_a-zA-Z0-9]+)\.(?:so|dll)$/', $file, $matches)) {
-                if (!extension_loaded($matches[1]) && @dl($matches[1])) {
-                    $exts[] = $matches[1];
+        if (is_dir($ext_dir)) {
+            foreach (scandir($ext_dir) as $file) {
+                if (preg_match('/^(?:php_)?([_a-zA-Z0-9]+)\.(?:so|dll)$/', $file, $matches)) {
+                    if (!extension_loaded($matches[1]) && @dl($matches[1])) {
+                        $exts[] = $matches[1];
+                    }
                 }
             }
         }
@@ -946,7 +948,7 @@ function save_results(string $output_file, bool $prompt_to_save_results): void
 
     $failed_tests_data .= "\n" . $sep . 'BUILD ENVIRONMENT' . $sep;
     $failed_tests_data .= "OS:\n" . PHP_OS . " - " . php_uname() . "\n\n";
-    $ldd = $autoconf = $sys_libtool = $libtool = $compiler = 'N/A';
+    $ldd = $autoconf = $sys_libtool = $libtool = $compiler = $cmake = 'N/A';
 
     if (!IS_WINDOWS) {
         /* If PHP_AUTOCONF is set, use it; otherwise, use 'autoconf'. */
@@ -957,7 +959,9 @@ function save_results(string $output_file, bool $prompt_to_save_results): void
         }
 
         /* Always use the generated libtool - Mac OSX uses 'glibtool' */
-        $libtool = shell_exec(INIT_DIR . '/libtool --version');
+        if (file_exists(INIT_DIR . '/libtool')) {
+            $libtool = shell_exec(INIT_DIR . '/libtool --version');
+        }
 
         /* Use shtool to find out if there is glibtool present (MacOSX) */
         $sys_libtool_path = shell_exec(__DIR__ . '/build/shtool path glibtool libtool');
@@ -979,9 +983,17 @@ function save_results(string $output_file, bool $prompt_to_save_results): void
         }
 
         $ldd = shell_exec("ldd $php 2>/dev/null");
+
+        /** If PHP_CMAKE is set, use it; otherwise, use 'cmake'. */
+        if (getenv('PHP_CMAKE')) {
+            $cmake = shell_exec(getenv('PHP_CMAKE') . ' --version');
+        } else {
+            $cmake = shell_exec('cmake --version');
+        }
     }
 
     $failed_tests_data .= "Autoconf:\n$autoconf\n";
+    $failed_tests_data .= "CMake:\n$cmake\n";
     $failed_tests_data .= "Bundled Libtool:\n$libtool\n";
     $failed_tests_data .= "System Libtool:\n$sys_libtool\n";
     $failed_tests_data .= "Compiler:\n$compiler\n";
